raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -584,211 +746,236 @@ private void updateVoterConnections(FindQuorumResponseData response) {
         }
     }
 
-    private boolean hasInconsistentLeader(int epoch, OptionalInt leaderId) {
+    private boolean hasConsistentLeader(int epoch, OptionalInt leaderId) {
         // Only elected leaders are sent in the request/response header, so if we have an elected
         // leaderId, it should be consistent with what is in the message.
-        if (epoch != quorum.epoch()) {
-            return false;
+        if (leaderId.isPresent() && leaderId.getAsInt() == quorum.localId) {
+            // The response indicates that we should be the leader, so we verify that is the case
+            return quorum.isLeader();
+        } else if (epoch != quorum.epoch()) {
+            return true;
         } else {
-            if (!quorum.leaderId().isPresent())
-                return false;
             if (!leaderId.isPresent())
-                return false;
-            return !quorum.leaderId().equals(leaderId);
-        }
-    }
-
-    private void becomeFollower(int epoch, OptionalInt leaderId) throws IOException {
-        if (leaderId.isPresent()) {
-            becomeFollower(leaderId.getAsInt(), epoch);
-        } else {
-            becomeUnattachedFollower(epoch);
+                return true;
+            if (!quorum.leaderId().isPresent())
+                return true;
+            return leaderId.equals(quorum.leaderId());
         }
     }
 
-    private boolean maybeHandleError(RaftResponse.Inbound response, long currentTimeMs) throws IOException {
-        ConnectionCache.ConnectionState connection = connections.getOrCreate(response.sourceId());
-
-        final ApiMessage data = response.data();
-        final Errors responseError;
-        final int responseEpoch;
-        final OptionalInt responseLeaderId;
-
-        if (data instanceof FetchQuorumRecordsResponseData) {
-            FetchQuorumRecordsResponseData fetchRecordsResponse = (FetchQuorumRecordsResponseData) data;
-            responseError = Errors.forCode(fetchRecordsResponse.errorCode());
-
-            if (responseError == Errors.OFFSET_OUT_OF_RANGE) {
-                connection.onResponseReceived(response.correlationId);
-                return false;
+    /**
+     * Handle response errors that are common across request types.
+     *
+     * @param error Error from the received response
+     * @param leaderId Optional leaderId from the response
+     * @param epoch Epoch received from the response
+     * @return Optional value indicating whether the error was handled here. True indicates
```

Comment:
I feel by listing:
```
Optional.empty
Optional(true)
Optional(false)
```
side by side with each case definition could achieve better clarity. So far for me I need to translate the optional first in mind and then understand the logic.